# Week 6 Data Visualization

### Setup

In [2]:
# Setup — run this cell first
# If any package is missing, it will install automatically
import subprocess, sys

def install(pkg):
    subprocess.check_call([sys.executable, "-m", "pip", "install", pkg, "-q"])

for pkg in ["pandas", "plotly", "kaleido", "nbformat"]:
    try:
        __import__(pkg)
    except ImportError:
        print(f"Installing {pkg}...")
        install(pkg)

import pandas as pd
import plotly.express as px

print("Setup complete.")

Setup complete.


In [1]:
!pip install plotly kaleido
fig.write_image("chart_name.png")


[notice] A new release of pip is available: 26.0.1 -> 26.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


NameError: name 'fig' is not defined

In [16]:
# Load your dataset

df = pd.read_csv("top_anime_100.csv")
rows, cols = df.shape
print(f"Rows: {rows}")
print(f"Columns: {cols}")

Rows: 100
Columns: 59


In [5]:
# Check column types and missing values
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 59 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   mal_id                            100 non-null    int64  
 1   url                               100 non-null    str    
 2   approved                          100 non-null    bool   
 3   titles                            100 non-null    str    
 4   title                             100 non-null    str    
 5   title_english                     95 non-null     str    
 6   title_japanese                    100 non-null    str    
 7   title_synonyms                    100 non-null    str    
 8   type                              100 non-null    str    
 9   source                            100 non-null    str    
 10  episodes                          98 non-null     float64
 11  status                            100 non-null    str    
 12  airing              

In [6]:
# Summary statistics for numeric columns
df.describe()

,mal_id,episodes,score,scored_by,rank,popularity,members,favorites,year,trailer.youtube_id,...,trailer.images.small_image_url,trailer.images.medium_image_url,trailer.images.large_image_url,trailer.images.maximum_image_url,aired.prop.from.day,aired.prop.from.month,aired.prop.from.year,aired.prop.to.day,aired.prop.to.month,aired.prop.to.year
count,100.000000,98.000000,100.000000,1.000000e+02,99.000000,100.000000,1.000000e+02,100.000000,69.000000,0.0,...,0.0,0.0,0.0,0.0,100.000000,100.000000,100.000000,75.000000,75.000000,75.000000
mean,35619.390000,19.846939,8.794500,4.965350e+05,50.121212,917.430000,8.640355e+05,30097.850000,2016.536232,NaN,...,NaN,NaN,NaN,NaN,11.150000,6.050000,2016.970000,20.666667,6.666667,2016.960000
std,21128.468762,28.702653,0.144176,5.608381e+05,28.920818,1172.400234,8.656511e+05,49603.626748,8.780962,NaN,...,NaN,NaN,NaN,NaN,7.375054,3.562926,9.074835,9.099054,3.457594,8.538371
min,1.000000,1.000000,8.620000,2.658000e+03,1.000000,3.000000,1.073100e+04,81.000000,1980.000000,NaN,...,NaN,NaN,NaN,NaN,1.000000,1.000000,1980.000000,1.000000,1.000000,1981.000000
25%,16659.750000,7.750000,8.687500,1.112462e+05,25.500000,113.250000,2.526785e+05,2553.750000,2012.000000,NaN,...,NaN,NaN,NaN,NaN,6.000000,4.000000,2013.000000,15.000000,3.000000,2013.000000
50%,39690.000000,13.000000,8.750000,2.291865e+05,50.000000,539.000000,4.835015e+05,9177.000000,2019.000000,NaN,...,NaN,NaN,NaN,NaN,8.000000,6.500000,2020.000000,24.000000,6.000000,2019.000000
75%,53279.000000,24.000000,8.900000,8.271682e+05,74.500000,1113.500000,1.340763e+06,31923.500000,2023.000000,NaN,...,NaN,NaN,NaN,NaN,16.250000,10.000000,2024.000000,28.000000,9.000000,2023.500000
max,61952.000000,201.000000,9.270000,2.310051e+06,102.000000,7076.000000,3.679834e+06,251747.000000,2026.000000,NaN,...,NaN,NaN,NaN,NaN,30.000000,12.000000,2026.000000,31.000000,12.000000,2026.000000


---

### **Question 1:** Among the 10 studios with the most titles in the top 100 anime, which has the highest average user score?

In [7]:
# Retrieve the top 10 studios with the highest average user score,
# then sort them from highest to lowest score
df.groupby('studios')['score'].mean().sort_values(ascending=False).head(10)

studios
[{'mal_id': 287, 'type': 'anime', 'name': 'David Production', 'url': 'https://myanimelist.net/anime/producer/287/David_Production'}]                                                                                                                                              9.140000
[{'mal_id': 1269, 'type': 'anime', 'name': 'K-Factory', 'url': 'https://myanimelist.net/anime/producer/1269/K-Factory'}, {'mal_id': 2256, 'type': 'anime', 'name': 'Kitty Film Mitaka Studio', 'url': 'https://myanimelist.net/anime/producer/2256/Kitty_Film_Mitaka_Studio'}]    9.020000
[{'mal_id': 314, 'type': 'anime', 'name': 'White Fox', 'url': 'https://myanimelist.net/anime/producer/314/White_Fox'}]                                                                                                                                                            8.995000
[{'mal_id': 1258, 'type': 'anime', 'name': 'Bandai Namco Pictures', 'url': 'https://myanimelist.net/anime/producer/1258/Bandai_Namco_Pictures'}

In [8]:
# Reorganize the data above into a tidy table for readability, while omitting mal_id, 
# type, anime, name, and URL as these details are irrelevant to the question

import ast
import pandas as pd
from IPython.display import display

def studio_names(cell):
    if pd.isna(cell):
        return []
    try:
        lst = ast.literal_eval(cell)
    except (ValueError, SyntaxError):
        return []
    return [s["name"] for s in lst]

tmp = df.assign(studio_name=df["studios"].apply(studio_names)).explode("studio_name")

top10_by_score = (
    tmp.groupby("studio_name", as_index=False)
    .agg(mean_score=("score", "mean"), title_count=("mal_id", "count"))
    .sort_values("mean_score", ascending=False)
    .head(10)
    .reset_index(drop=True)
)

display(top10_by_score)

,studio_name,mean_score,title_count
0,David Production,9.140000,1
1,K-Factory,9.020000,1
2,Kitty Film Mitaka Studio,9.020000,1
3,White Fox,8.995000,2
4,Bandai Namco Pictures,8.954000,5
5,TMS Entertainment,8.930000,1
6,Madhouse,8.894286,7
7,Sunrise,8.892857,7
8,TOHO animation STUDIO,8.880000,2
9,Bones,8.870000,3


**Note:**
After looking at the data in the above table, I realized that the suggested code was not correctly answering my question. My question asked, “Among the top 10 studios with the most titles...,” but it seems the Cursor agent only sorted studios by highest average user score rather than accounting for the number of titles produced by each studio.

In [9]:
# Your analysis for Question 1
# First find studios with the most titles, then compare average user score

import ast
import pandas as pd
from IPython.display import display

# Give me a simple list of studio names
def studio_names(cell):
    if pd.isna(cell):
        return []
    try:
        lst = ast.literal_eval(cell)
    except (ValueError, SyntaxError):
        return []
    return [s["name"] for s in lst]

# Turn that list into 1 row per studio, so I can count and average by studio
tmp = df.assign(studio_name=df["studios"].apply(studio_names)).explode("studio_name")

# Summarize every studio with average score and number of titles
studio_stats = (
    tmp.groupby("studio_name", as_index=False)
    .agg(mean_score=("score", "mean"), title_count=("mal_id", "count"))
)

# Collect the top 10 studios with the most titles and their average scores
top10_most_titles = (
    studio_stats.sort_values(["title_count", "mean_score"], ascending=[False, False])
    .head(10)
    .reset_index(drop=True)
)

display(top10_most_titles)

# Show the studio that has the highest average score, which answers my question
best_avg_in_top10 = top10_most_titles.sort_values("mean_score", ascending=False).head(1)
display(best_avg_in_top10)

,studio_name,mean_score,title_count
0,MAPPA,8.811250,8
1,Madhouse,8.894286,7
2,Sunrise,8.892857,7
3,Studio Pierrot,8.761429,7
4,Bandai Namco Pictures,8.954000,5
5,Kyoto Animation,8.822000,5
6,Shaft,8.790000,5
7,A-1 Pictures,8.760000,4
8,Studio Signpost,8.742500,4
9,Bones,8.870000,3


,studio_name,mean_score,title_count
4,Bandai Namco Pictures,8.954,5


 **Answer: Out of the top 10 studios with the most titles in the top 100 anime, Bandai Namco Pictures has the highest average score of 8.954, with 5 titles represented.**

In [35]:
import plotly.express as px

chart_q1 = top10_most_titles.sort_values("mean_score", ascending=True)
chart_q1["bar_label"] = (
    chart_q1["mean_score"].round(4).astype(str)
    + "  (" + chart_q1["title_count"].astype(int).astype(str) + " titles)"
)

def wrap_name(name, max_len=12):
    if len(name) <= max_len:
        return name
    words = name.split(" ")
    lines, current = [], words[0]
    for w in words[1:]:
        if len(current) + 1 + len(w) > max_len:
            lines.append(current)
            current = w
        else:
            current += " " + w
    lines.append(current)
    return "<br>".join(lines)

chart_q1["studio_wrapped"] = chart_q1["studio_name"].apply(wrap_name)

fig = px.bar(
    chart_q1,
    x="mean_score",
    y="studio_wrapped",
    orientation="h",
    title="Highest Average Score by Top 10 Studios with the Most Titles in the Top 100 Anime",
    labels={"studio_wrapped": "Studio", "mean_score": "Average Score"},
    text="bar_label",
)

fig.update_xaxes(range=[8.6, 9.1])
fig.update_traces(textposition="outside")
fig.update_layout(
    yaxis_title="Studio",
    xaxis_title="Average Score",
    margin=dict(l=160, r=140),
    height=550, width=800,
)
fig.show()

---

### **Question 2:** *Among the top 100 anime, which 5 genres appear most frequently?*

In [18]:
# Your analysis for Question 2
import ast
from IPython.display import display

# Produce a clean list of genre names
def genre_names(cell):
    if pd.isna(cell):
        return []
    try:
        lst = ast.literal_eval(cell)
    except (ValueError, SyntaxError):
        return []
    return [g["name"] for g in lst]

# Tell me which 5 genres occur most often and how many times each occur
top5_genre_names = (
    df.assign(genre_name=df["genres"].apply(genre_names))
    .explode("genre_name")
    ["genre_name"]
    .value_counts()
    .head(5)
    .rename_axis("genre")
    .reset_index(name="count")
)

display(top5_genre_names)

,genre,count
0,Drama,47
1,Action,44
2,Adventure,22
3,Supernatural,19
4,Comedy,19


**Answer: The 5 genres that appear most frequently in the top 100 anime are: Drama (47), Action (44), Adventure (22), Supernatural (19), and Comedy (19).**

In [21]:
import plotly.express as px

fig = px.bar(
    top5_genre_names,
    x="genre",
    y="count",
    title="Top 5 Genres from the Top 100 Anime",
    labels={"genre": "Genre", "count": "Number of Titles"},
    text="count",
)

fig.update_traces(textposition="outside")
fig.update_layout(xaxis_title="Genre", yaxis_title="Number of Titles")
fig.show()

---

### **Question 3:** Among the top 100 anime, which decade of release (e.g., 2000s, 2010s, 2020s) has the highest average user score?

In [10]:
import pandas as pd
from IPython.display import display

# Keep anime with a known year in 1980-2026
period = df.dropna(subset=["year"]).loc[lambda d: (d["year"] >= 1980) & (d["year"] <= 2026)].copy()

# Bucket anime into 5 decade ranges
period["decade"] = pd.cut(
    period["year"],
    bins=[1979, 1989, 1999, 2009, 2019, 2026],
    labels=["1980-1989", "1990-1999", "2000-2009", "2010-2019", "2020-2026"],
    include_lowest=True,
    right=True,
)

# Compute average score and count per range, and sort by average score
score_by_decade = (
    period.groupby("decade", observed=True)
    .agg(mean_score=("score", "mean"), anime_count=("mal_id", "count"))
    .reset_index()
    .sort_values("mean_score", ascending=False)
)

display(score_by_decade)

# Report the top decade
winner = score_by_decade.iloc[0]
print(
    f"Highest average score: {winner['decade']} "
    f"(mean score {winner['mean_score']:.4f}, n={int(winner['anime_count'])})"
)

,decade,mean_score,anime_count
2,2000-2009,8.823000,10
3,2010-2019,8.820435,23
0,1980-1989,8.820000,1
4,2020-2026,8.782500,32
1,1990-1999,8.720000,3


Highest average score: 2000-2009 (mean score 8.8230, n=10)


**Note:** When I reviewed the table and added the counts, I realized they only summed to 69, even though there should have been a total of 100 anime. I then asked Cursor where the other 31 entries went, and it revealed that those anime did not have a year listed.

In [11]:
# Your analysis for Question 3
# Missing year -> "Unknown". Still excludes non-null years before 1980 or after 2026.

# Check through titles to see if they have a listed year between 1980 and 2026
has_year = df["year"].notna()
in_range = (df["year"] >= 1980) & (df["year"] <= 2026)

# For animes with known years, categorize them into decade range
bucketed = df.loc[has_year & in_range].copy()
bucketed["decade"] = pd.cut(
    bucketed["year"],
    bins=[1979, 1989, 1999, 2009, 2019, 2026],
    labels=["1980-1989", "1990-1999", "2000-2009", "2010-2019", "2020-2026"],
    include_lowest=True,
    right=True,
)

# Take anime with unlisted year and put them together in Unknown group
unknown_year = df.loc[~has_year].copy()
unknown_year["decade"] = "Unknown"

period_with_unknown = pd.concat([bucketed, unknown_year], ignore_index=True)

# Compute average score and count per range, and sort by average score
score_by_decade_with_unknown = (
    period_with_unknown.groupby("decade", observed=True)
    .agg(mean_score=("score", "mean"), anime_count=("mal_id", "count"))
    .reset_index()
    .sort_values("mean_score", ascending=False)
)

display(score_by_decade_with_unknown)

# Display decade with top average score
winner_u = score_by_decade_with_unknown.iloc[0]
print(
    f"Highest average score: {winner_u['decade']} "
    f"(mean score {winner_u['mean_score']:.4f}, n={int(winner_u['anime_count'])})"
)

,decade,mean_score,anime_count
2,2000-2009,8.823000,10
3,2010-2019,8.820435,23
0,1980-1989,8.820000,1
5,Unknown,8.784839,31
4,2020-2026,8.782500,32
1,1990-1999,8.720000,3


Highest average score: 2000-2009 (mean score 8.8230, n=10)


**Answer: The decade with the highest average score is 2000–2009, with an average score of 8.8230 across 10 anime titles.**

In [14]:
import plotly.express as px

decade_labels = {
    "1980-1989": "1980's<br><sub>(1980-1989)</sub>",
    "1990-1999": "1990's<br><sub>(1990-1999)</sub>",
    "2000-2009": "2000's<br><sub>(2000-2009)</sub>",
    "2010-2019": "2010's<br><sub>(2010-2019)</sub>",
    "2020-2026": "2020's<br><sub>(2020-2026)</sub>",
    "Unknown": "Unknown",
}

chart_df = score_by_decade_with_unknown.copy()
chart_df["decade_label"] = chart_df["decade"].map(decade_labels)
chart_df = chart_df.sort_values("mean_score", ascending=True)
chart_df["bar_label"] = (
    chart_df["mean_score"].round(4).astype(str)
    + "  (" + chart_df["anime_count"].astype(int).astype(str) + " titles)"
)

fig = px.bar(
    chart_df,
    x="mean_score",
    y="decade_label",
    orientation="h",
    title="Highest Average Score by Decade for Top 100 Anime",
    labels={"decade_label": "Decade", "mean_score": "Average Score"},
    text="bar_label",
)

fig.update_xaxes(range=[8.6, 8.9])
fig.update_traces(textposition="outside")
fig.update_layout(yaxis_title="Decade", xaxis_title="Average Score", margin=dict(l=120))
fig.show()